### PR TITLE
InfiAgent-DABench

### PR DESCRIPTION
**Pull request by kithib**
**Features**
This PR is used to solve the InfiAgent-DABench using Data Interpreter (DI), and obtains 94.93% accuracy using gpt-4o

**Feature Docs**
detailed in README.md

**Result**
截屏2024-10-28 21 59 38

### CODE REVIEW

File: examples/di/InfiAgent-DABench/DABench.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    def get_prompt(self, question_id):
+        """Retrieve the question by its id."""
+        temp = self.get_question(question_id)
+        return self.template.format(question=temp['question'], constraints=temp['constraints'], format=temp['format'], file_name= str(DABENCH_PATH) + '/da-dev-tables/' + temp['file_name'], level=temp['level'],)
+
+    def get_answer(self, answer_id):
+        """Retrieve the answer list by its id."""
+        return self.answers.get(answer_id, "Answer not found.")
+
+    def eval(self, id, prediction):
```
Comment (stellaHSR) [inline examples/di/InfiAgent-DABench/DABench.py:29]: Should add the reference : https://github.com/InfiAgent/InfiAgent/blob/main/examples/DA-Agent/eval_closed_form.py

File: examples/di/InfiAgent-DABench/DABench.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
+import json
+from pathlib import Path
+from metagpt.const import DABENCH_PATH
+from examples.di.requirements_prompt import DABENCH
+class DABench:
```
Comment (stellaHSR) [inline examples/di/InfiAgent-DABench/DABench.py:5]: 在class和import之间增加空行，遵循PEP8

**Comment by stellaHSR** [comment]
Thanks. Some pre-commit checks didn't pass. Please install pre-commit locally and rerun the checks before submitting again.

File: examples/di/InfiAgent-DABench/DABench.py (hunk 1 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/DABench.py:29]: ok

File: examples/di/InfiAgent-DABench/DABench.py (hunk 2 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/DABench.py:5]: ok

File: examples/di/InfiAgent-DABench/DABench.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+    return round(correct / total, 4) if total > 0 else 0
+
+
+def evaluate_accuracy_by_sub_question(results):
+    correct = sum(sum(result["correctness"].values()) for result in results if "correctness" in result)
+    total = sum(len(result["correctness"]) for result in results if "correctness" in result)
+    return round(correct / total, 4) if total > 0 else 0
+
+
+def evaluate_accuracy_proportional_by_sub_question_adjusted(results):
```
Comment (stellaHSR) [inline examples/di/InfiAgent-DABench/DABench.py:21]: 这三个函数最好补充下简单的注释，比如对应其榜单上的维度

**Comment by kithib** [comment]
> Thanks. Some pre-commit checks didn't pass. Please install pre-commit locally and rerun the checks before submitting again.
ok

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+from DABench import DABench
+
+from metagpt.roles.di.data_interpreter import DataInterpreter
+
+
+async def main():
+    """Evaluate all"""
+    DA = DABench()
+    id_list, predictions = [], []
+    for key, value in DA.answers.items():
```
Comment (stellaHSR) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py:13]: 这里建议后续提供一个并行的入口，增加并发，考虑使用`predictions = await asyncio.gather(*tasks)`；

File: metagpt/const.py [hunk 5]
```diff
@@ -43,6 +43,7 @@ def get_metagpt_root():
 EXAMPLE_PATH = METAGPT_ROOT / "examples"
 EXAMPLE_DATA_PATH = EXAMPLE_PATH / "data"
 DATA_PATH = METAGPT_ROOT / "data"
+DABENCH_PATH = EXAMPLE_PATH / "di/InfiAgent-DABench/InfiAgent/examples/DA-Agent/data"
```
Comment (mannaandpoem) [inline metagpt/const.py:46]: 这个路径名称是否有点重复，DABENCH_PATH 中出现了两个 examples

File: examples/di/InfiAgent-DABench/DABench.py [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+class DABench:
+    def __init__(
+        self,
+        questions_file=Path(DABENCH_PATH) / "da-dev-questions.jsonl",
+        answers_file=Path(DABENCH_PATH) / "da-dev-labels.jsonl",
+        template="",
+    ):
+        # Read questions from a JSONL file
+        with open(questions_file, "r") as file:
+            self.questions = {int(json.loads(line)["id"]): json.loads(line) for line in file}
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/DABench.py:41]: It is recommended to use specific functions to save/load jsonl. If you can't find one, write one.

File: examples/di/InfiAgent-DABench/DABench.py [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+            if isinstance(true_value, str) and (metric not in pred_dict or str(pred_dict[metric]) != str(true_value)):
+                correct = False
+                break
+
+        return correct
+
+    def eval_all(self, id_list, predictions):
+        """Evaluate all predictions and calculate accuracy rates."""
+
+        def sigle_eval(id, prediction):
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/DABench.py:109]: typo

File: examples/di/InfiAgent-DABench/DABench.py [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+        # Sort the true labels to match the order of predictions
+        sorted_true_label = sorted(true_label, key=lambda x: x[0])
+
+        # Compare each prediction with the corresponding true label
+        correct = True
+        for metric, true_value in sorted_true_label:
+            try:
+                true_value = float(true_value)
+            except:
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/DABench.py:93]: Used too many try / except. But this is not our style.

File: examples/di/InfiAgent-DABench/DABench.py [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+        self.template = template if template else DABENCH
+
+    def get_question(self, question_id):
+        """Retrieve the question by its id."""
+        return self.questions.get(question_id, "Question not found.")
+
+    def get_prompt(self, question_id):
+        """Retrieve the question by its id."""
+        temp = self.get_question(question_id)
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/DABench.py:55]: This name doesn't make much sense. Can you change it to a more meaningful one?

File: examples/di/InfiAgent-DABench/README.md [hunk 10]
```diff
@@ -0,0 +1,13 @@
+# InfiAgent-DABench
+This example is used to solve the InfiAgent-DABench using Data Interpreter (DI), and obtains 94.93% accuracy using gpt-4o.
+
+## Dataset-install
+```
+git clone https://github.com/InfiAgent/InfiAgent.git
+```
+## How to run
+```
+python run_InfiAgent-DABench_sigle.py --id x   # run a task
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/README.md:10]: typo

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+        prediction = prediction_json[-1]["result"]
+        return prediction
+    except Exception as e:
+        print(f"Error processing requirement: {requirement}. Error: {e}")
+        return None
+
+
+async def evaluate_all(agent_class):
+    """Evaluate all tasks in DABench using the specified baseline agent"""
+    DA = DABench()
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py:30]: Variable naming should comply with python specifications

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+            predictions.append(str(prediction))
+            labels.append(str(DA.get_answer(key)))
+        except:
+            id_list.append(key)
+            is_true.append(str(DA.eval(key, "")))
+            predictions.append(str(""))
+            labels.append(str(DA.get_answer(key)))
+    df = pd.DataFrame({"Label": labels, "Prediction": predictions, "T/F": is_true})
+
+    # 将DataFrame写入Excel文件
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py:31]: Use English comments

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+    df.to_excel("output.xlsx", index=False)
+    print(DA.eval_all(id_list, predictions))
+    # 将列表转换为pandas DataFrame
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py:34]: ditto

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+
+async def main(id=0):
+    DA = DABench()
+    requirement = DA.get_prompt(id)
+    di = DataInterpreter()
+    result = await di.run(requirement)
+    prediction = json.loads(str(result).split("Current Plan")[1].split("## Current Task")[0])[-1]["result"]
+    is_correct = DA.eval(id, prediction)
+    print(f"Prediction is {'correct' if is_correct else 'incorrect'}.")
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py:16]: I recommend using metagpt logger.

File: examples/di/InfiAgent-DABench/DABench.py (hunk 3 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/DABench.py:25]: done

File: examples/di/InfiAgent-DABench/DABench.py (hunk 6 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/DABench.py:45]: done

File: examples/di/InfiAgent-DABench/DABench.py (hunk 7 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/DABench.py:108]: done

File: examples/di/InfiAgent-DABench/DABench.py (hunk 8 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/DABench.py:94]: done

File: examples/di/InfiAgent-DABench/DABench.py (hunk 9 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/DABench.py:59]: done

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py (hunk 14 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py:17]: done

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py (hunk 13 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py]: done

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py (hunk 12 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py]: done

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py (hunk 4 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py:15]: [run_InfiAgent-DABench_all.py] 串行测试DABench
[run_InfiAgent-DABench.py] 并行测试DABench，新修改的代码增加了参数k，可以k个一组进行并行

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py (hunk 11 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py:30]: done

File: metagpt/const.py (hunk 5 above)
Comment (kithib) [reply metagpt/const.py]: done

File: examples/di/InfiAgent-DABench/README.md (hunk 10 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/README.md]: done

File: examples/di/InfiAgent-DABench/README.md [hunk 15]
```diff
@@ -0,0 +1,15 @@
...
+
+## Dataset
+```
+cd /examples/di/InfiAgent-DABench
+git clone https://github.com/InfiAgent/InfiAgent.git
+mv InfiAgent/examples/DA-Agent/data ./
+```
+## How to run
+```
+python run_InfiAgent-DABench_sigle.py --id x   # run a task, x represents the id of the question you want to test
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/README.md:12]: typo

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+    for i in range(0, len(tasks), k):
+        # Get the current group of tasks
+        current_group = tasks[i : i + k]
+        # Execute the current group of tasks in parallel
+        group_predictions = await asyncio.gather(*current_group)
+        # Filter out any None values from the predictions and extend the predictions list
+        predictions.extend(pred for pred in group_predictions if pred is not None)
+
+    # Evaluate the results using all valid predictions and print the evaluation
+    print(DA.eval_all(id_list, predictions))
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py:66]: use MetaGPT logger instead.

File: examples/di/InfiAgent-DABench/DABench.py [hunk 17]
```diff
@@ -0,0 +1,486 @@
...
+            "accuracy_by_sub_question": accuracy_by_sub_question,
+            "proportional_accuracy_by_sub_question": proportional_accuracy_by_sub_question,
+        }
+
+
+if __name__ == "__main__":
+    DA = DABench()
+    id = 0
+    prediction = "@mean_fare[34.65]"
+    print(DA.eval(id, prediction))
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/DABench.py:479]: use MetaGPT logger instead.

File: examples/di/InfiAgent-DABench/DABench.py [hunk 18]
```diff
@@ -0,0 +1,486 @@
...
+    ids = [0, 5, 6]
+    predictions = [
+        "@mean_fare[34.89]",
+        "@correlation_coefficient[0.21]",
+        "@mean_fare_child[31.09], @mean_fare_teenager[31.98], @mean_fare_adult[35.17], @mean_fare_elderly[43.47]",
+    ]
+    print(DA.eval_all(ids, predictions))
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/DABench.py:486]: use MetaGPT logger instead.

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+        result = await agent.run(requirement)
+
+        # Parse the result to extract the prediction from the JSON response
+        prediction_json = json.loads(str(result).split("Current Plan")[1].split("## Current Task")[0])
+        prediction = prediction_json[-1]["result"]  # Extract the last result from the parsed JSON
+
+        return prediction  # Return the extracted prediction
+    except Exception as e:
+        # Log an error message if an exception occurs during processing
+        print(f"Error processing requirement: {requirement}. Error: {e}")
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py:33]: use MetaGPT logger instead.

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+            save_history(role=di)
+            temp_prediction, temp_istrue = DA.eval(key, str(result))
+            is_true.append(str(temp_istrue))
+            predictions.append(str(temp_prediction))
+        except:
+            is_true.append(str(DA.eval(key, "")))
+            predictions.append(str(""))
+    df = pd.DataFrame({"Label": labels, "Prediction": predictions, "T/F": is_true})
+    df.to_excel("DABench_output.xlsx", index=False)
+    print(DA.eval_all(id_list, predictions))
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py:31]: use MetaGPT logger instead.

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py (hunk 14 above)
Comment (geekan) [reply examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py:18]: not done yet?

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py [hunk 21]
```diff
@@ -0,0 +1,22 @@
+import fire
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py:1]: should this filename be renamed to *_single.py ?

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+from DABench import DABench
+
+from metagpt.logs import logger
+from metagpt.roles.di.data_interpreter import DataInterpreter
+from metagpt.utils.recovery_util import save_history
+
+
+async def main(id=0):
+    """Evaluate one task"""
+    DA = DABench()
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py:11]: I prefer to use **bench** as a variable. Generally speaking, python PEP8 does not allow uppercase variable names

File: examples/di/InfiAgent-DABench/DABench.py [hunk 23]
```diff
@@ -0,0 +1,486 @@
...
+
+        return {
+            "accuracy_by_question": accuracy_by_question,
+            "accuracy_by_sub_question": accuracy_by_sub_question,
+            "proportional_accuracy_by_sub_question": proportional_accuracy_by_sub_question,
+        }
+
+
+if __name__ == "__main__":
+    DA = DABench()
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/DABench.py:476]: I prefer to use **bench** as a variable. Generally speaking, python PEP8 does not allow uppercase variable names

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+async def evaluate_all(agent, k):
+    """Evaluate all tasks in DABench using the specified baseline agent.
+
+    Tasks are divided into groups of size k and processed in parallel.
+
+    Args:
+        agent: The baseline agent used for making predictions.
+        k (int): The number of tasks to process in each group concurrently.
+    """
+    DA = DABench()  # Create an instance of DABench to access its methods and data
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py:46]: I prefer to use **bench** as a variable. Generally speaking, python PEP8 does not allow uppercase variable names

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py [hunk 25]
```diff
@@ -0,0 +1,35 @@
...
+from DABench import DABench
+
+from metagpt.logs import logger
+from metagpt.roles.di.data_interpreter import DataInterpreter
+from metagpt.utils.recovery_util import save_history
+
+
+async def main():
+    """Evaluate all"""
+    DA = DABench()
```
Comment (geekan) [inline examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py:12]: I prefer to use **bench** as a variable. Generally speaking, python PEP8 does not allow uppercase variable names

**Comment by geekan** [comment]
BTW, the picture is not displayed properly.

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py (hunk 14 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py]: sorry, I have solved this problem now.

File: examples/di/InfiAgent-DABench/DABench.py (hunk 18 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/DABench.py]: done

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py (hunk 20 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py]: done

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py (hunk 22 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/run_InfiAgent-DABench_sigle.py]: done

File: examples/di/InfiAgent-DABench/DABench.py (hunk 23 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/DABench.py]: done

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py (hunk 24 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/run_InfiAgent-DABench.py]: done

File: examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py (hunk 25 above)
Comment (kithib) [reply examples/di/InfiAgent-DABench/run_InfiAgent-DABench_all.py]: done

**Comment by kithib** [comment]
> BTW, the picture is not displayed properly.

截屏2024-10-28 21 59 38

**Comment by stellaHSR** [comment]
LGTM

**Review by geekan — APPROVED** [review]
LGTM

**Pull request merged**